src/bluesearch/database/identifiers.py

```diff
@@ -0,0 +1,80 @@
+"""Module for handling identifiers."""
+
+# Blue Brain Search is a text mining toolbox focused on scientific use cases.
+#
+# Copyright (C) 2020  Blue Brain Project, EPFL.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+import hashlib
+from typing import List, Tuple
+
+import pandas as pd
+
+
+def generate_uid(identifiers: Tuple) -> str:
+    """Generate a deterministic UID for the given identifiers.
+
+    Parameters
+    ----------
+    identifiers
+        Values of the identifiers.
+
+    Returns
+    -------
+    int
+        A deterministic UID.
+    """
+    data = str(identifiers).encode()
+    hashed = hashlib.md5(data).hexdigest()  # nosec
+    return hashed
+
+
+def generate_uids(metadata: pd.DataFrame, identifiers: List[str]) -> pd.DataFrame:
+    """Generate UIDs for papers with multiple sources, identifiers.
+
+    Papers with the same values for the given identifiers get the same UID.
+
+    Missing values should have the value 'None', which is considered a value by itself.
+    Then, identifiers (a, None) and identifiers (a, b) have two different UIDs.
+    Papers with all given identifiers unspecified have 'None' as UID.
+
+    The generation of the UID is deterministic.
+
+    Parameters
+    ----------
+    metadata
+        Paper metadata including the given identifiers.
+    identifiers
+        Columns of the identifiers.
+
+    Returns
+    -------
+    pandas.DataFrame
+        Paper metadata with a new column with the generated UIDs.
+    """
+    # Ignore papers without values for the given identifiers.
+
+    df = metadata.dropna(how="all", subset=identifiers)
```

Comment:
This is a safeguard to prevent an unexpected behaviour. Indeed, papers without identifiers would be otherwise given the same UID. Different papers would be considered identical while they are not. This could happen when the parsing of the identifiers fails, for example.

Reply:
Ok, I am also wondering whether one of the below solutions wouldn't make more sense
1) Raising an exception if at least one paper has all identifiers equal to `None`
2) Let `generate_uid` output `None` if all identifiers are `None`

The reason why I am mentioning this is that this function can effectively kick out some articles (`len(df) <= len(metadata)`) without letting the user know and it might be annoying to try to figure out the diff after the fact.

Reply:
> this function can effectively kick out some articles

Oh. Actually, `metadata["uid"] = df[identifiers].apply(_uid, axis=1)` wasn't a refactoring issue. The goal was indeed to assign `None` for `uid` for the articles with all identifiers empty. Anyway, the next comment solves this.

> Let `generate_uid` output `None` if all identifiers are `None`.

That's the case in 26bea59.

Reply:
Great. Thank you!